components/repeater/index.js

```diff
@@ -0,0 +1,256 @@
+import { v4 as uuid } from 'uuid';
+
+import {
+	DndContext,
+	closestCenter,
+	KeyboardSensor,
+	PointerSensor,
+	useSensor,
+	useSensors,
+} from '@dnd-kit/core';
+import {
+	arrayMove,
+	SortableContext,
+	sortableKeyboardCoordinates,
+	verticalListSortingStrategy,
+	useSortable,
+} from '@dnd-kit/sortable';
+import { restrictToVerticalAxis } from '@dnd-kit/modifiers';
+import { CSS } from '@dnd-kit/utilities';
+import { DragHandle } from '../content-picker/PickedItem';
+
+/**
+ * The Repeater Component.
+ *
+ * @param {object} props React props
+ * @param {Function} props.children Render prop to render the children.
+ * @param {string} props.attribute property of the block attribute that will provide data for Repeater.
+ * @param {string} props.addButton render prop to customize the "Add item" button.
+ * @param {boolean} props.allowReordering boolean to toggle reordering of Repeater items.
+ * @returns {*} React JSX
+ */
+export const Repeater = ({ children, attribute, addButton, allowReordering }) => {
+	const { clientId, name } = useBlockEditContext();
+	const { updateBlockAttributes } = dispatch(blockEditorStore);
+
+	const sensors = useSensors(
+		useSensor(PointerSensor),
+		useSensor(KeyboardSensor, {
+			coordinateGetter: sortableKeyboardCoordinates,
+		}),
+	);
+
+	const { repeaterData, defaultRepeaterData } = useSelect((select) => {
+		const { getBlockAttributes } = select(blockEditorStore);
+		const { getBlockType } = select(blocksStore);
+		const repeaterDataTemp = getBlockAttributes(clientId)[attribute];
+
+		if (repeaterDataTemp.length === 1 && !repeaterDataTemp[0].id) {
+			repeaterDataTemp[0].id = uuid();
+		}
+
+		return {
+			repeaterData: repeaterDataTemp,
+			defaultRepeaterData: getBlockType(name).attributes[attribute].default,
+		};
+	});
+
+	function handleDragEnd(event) {
+		const { active, over } = event;
+
+		if (active.id !== over.id) {
+			const moveArray = (items) => {
+				const oldIndex = items.findIndex((item) => item.id === active.id);
+				const newIndex = items.findIndex((item) => item.id === over.id);
+
+				return arrayMove(items, oldIndex, newIndex);
+			};
+
+			updateBlockAttributes(clientId, {
+				[attribute]: moveArray(repeaterData),
+			});
+		}
+	}
+
+	/**
+	 * Adds a new repeater item.
+	 */
+	function addItem() {
+		const defaultRepeaterDataCopy = JSON.parse(JSON.stringify(defaultRepeaterData));
+
+		if (!defaultRepeaterData.length) {
+			defaultRepeaterDataCopy.push([]);
+		}
+
+		defaultRepeaterDataCopy[0].id = uuid();
+
+		updateBlockAttributes(clientId, {
+			[attribute]: [...repeaterData, ...defaultRepeaterDataCopy],
+		});
+	}
+
+	/**
+	 * Updates the item currently being edited.
+	 *
+	 * @param {string|number|boolean} value The value that should be used to updated the item.
+	 * @param {number} index The index at which the item should be updated.
+	 */
+	function setItem(value, index) {
+		const repeaterDataCopy = JSON.parse(JSON.stringify(repeaterData));
```

Comment:
Let's use spreading here since it is a more common pattern. 

```suggestion
		const repeaterDataCopy = [...repeaterData];
```

Reply:
@fabiankaegy I made this change because the earlier version caused problems due to shallow-copy.
Any changes made to the objects in the `repeaterDataCopy` array were also reflected in `repeaterData`.
What is your opinion on using [structeredClone()](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone)?

Reply:
@Sidsector9 okay that makes sense then :) maybe just add a code comment on here so it is easier to understand why this was implemented that way so it is easier to maintain :)